### ISSUE TITLE
Auxiliary task (different type)

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/snowfall#97]
This is an auxiliary task for use with MMI training, as an alternative to the "attention" auxiliary task, that will be easier to do.  It is quite similar to the xent regularization in Kaldi.

From `num.get_arc_post().exp()` we can get the arc-level posteriors.  We can then construct a tensor `num_post_mat` of alignment posteriors of shape (seqframe_idx_dim, nnet_output_dim), where seqframe_idx_dim == dense_fsa_vec.shape_dim0() and nnet_output_dim is the num-phones plus one.  We use some kind of index_add_() construct to add in the posteriors to it; this will require `num.label` and `num.seqframe_idx`, assuming we gave `seqframe_idx_name='seqframe_idx'` to intersect_dense() for num.   Or use a torch sparse matrix.  Remember we have to eliminate elements where the labels are minus one, first, to avoid invalid indexing with `label < 0`.

Next, we need to give the neural network an auxiliary output layer that is structurally the same as the regular output, but with different parameters (it might be best to just have it return a single output that we divide into two pieces?).  The main extra part of the objective function would be  `xent_like = (xent_output * num_post_mat).sum()`, and note that xent_output needs to have been through a log-softmax or this is meaningless, i.e. it needs to be normalized logprobs.  From that I want to subtract the (appropriately scaled) negentropy of the distribution, to avoid the network being encouraged to put all the probability on one class.   Something like:  `phone_counts = num_post_mat.sum(0); xent_like = (nnet_output.log() * num_post_mat).sum() - (phone_counts * torch.log(phone_counts/phone_counts.sum() + 1.0e-20)).sum()`

**Comment by danpovey** [comment]
@zhu-han if you could work on this it would be great.

**Comment by zhu-han** [comment]
Working on it.

**Comment by danpovey** [comment]
Great!

**Comment by danpovey** [comment]
My understanding is that we tried this and it did not work.. closing the issue.
